### PR TITLE
Add support for 3d shapes in crop window

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

#### Why we need this PR?
*Pick one*
- Need to support n-dimensional coordinates in CropWindow and CropWindowGenerator (Slice family operators)
- First step before adding support for volumetric cropping windows in `Crop` operator

#### What happened in this PR?
 - What was changed, added, removed?
Changed using `x`, `y`, `w` and `h` for `anchor` and `shape`, represented as `kernels::TensorShape<>`
Adjusted usage of CropWindow generator to use the new API
 - What is most important part that reviewers should focus on?
The correctness of the changes from explicitly named coordinates to index based
 - Was this PR tested? How? 
Unit tests and python tests
 - Were docs and examples updated, if necessary?
NA
**JIRA TASK**: [DALI-1032]

### CODE REVIEW

File: dali/image/image.h [hunk 1]
```diff
@@ -79,15 +79,15 @@ class Image {
...
+    crop_window_generator_ = [crop_window](const kernels::TensorShape<>& shape) {
+      DALI_ENFORCE(crop_window.IsInRange(shape),
         "crop_window["
-        + std::to_string(crop_window.x)
-        + ", " + std::to_string(crop_window.y)
-        + ", " + std::to_string(crop_window.w)
-        + ", " + std::to_string(crop_window.h) + "]"
+        + std::to_string(crop_window.anchor[1])
+        + ", " + std::to_string(crop_window.anchor[0])
+        + ", " + std::to_string(crop_window.anchor[1])
```
Comment (jantonguirao) [inline dali/image/image.h]: should be `shape`

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

File: dali/util/crop_window.h [hunk 2]
```diff
@@ -16,48 +16,52 @@
...
-          && y >= 0
-          && y < H
-          && x+w >= 0
-          && x+w <= W
-          && y+h >= 0
-          && y+h <= H;
+    inline bool IsInRange(const kernels::TensorShape<>& input_shape) const {
+      bool res = true;
+      for (int dim = 0; dim < input_shape.size(); dim++)
+        res = res && anchor[dim] >= 0 && anchor[dim] + shape[dim] <= input_shape[dim];
```
Comment (mzient) [inline dali/util/crop_window.h]:
```suggestion
        if (anchor[dim] < 0 || anchor[dim] + shape[dim] > input_shape[dim])
          return false;
```

File: dali/util/crop_window.h [hunk 3]
```diff
@@ -16,48 +16,52 @@
...
+      return res;
```
Comment (mzient) [inline dali/util/crop_window.h]:
```suggestion
      return true;
```

File: dali/util/random_crop_generator.cc [hunk 4]
```diff
@@ -32,8 +32,9 @@ RandomCropGenerator::RandomCropGenerator(
   , num_attempts_(num_attempts) {
 }
 
-CropWindow RandomCropGenerator::GenerateCropWindowImpl(int H, int W) {
-  CropWindow crop = {};
+CropWindow RandomCropGenerator::GenerateCropWindowImpl(const kernels::TensorShape<>& shape) {
+  CropWindow crop;
+  int H = shape[0], W = shape[1];
```
Comment (mzient) [inline dali/util/random_crop_generator.cc]: Check `shape.size() == 2` first.

File: dali/util/crop_window.h (hunk 2 above)
Comment (jantonguirao) [reply dali/util/crop_window.h]: will do

File: dali/util/crop_window.h (hunk 3 above)
Comment (jantonguirao) [reply dali/util/crop_window.h]: will do

File: dali/util/random_crop_generator.cc (hunk 4 above)
Comment (jantonguirao) [reply dali/util/random_crop_generator.cc]: will do

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

File: dali/pipeline/operators/crop/crop_attr.h [hunk 5]
```diff
@@ -36,43 +36,55 @@ class CropAttr {
...
-      DALI_ENFORCE(!has_crop_h_arg && !has_crop_w_arg,
-        "`crop` argument is not compatible with `crop_h`, `crop_w`");
+      DALI_ENFORCE(!has_crop_h_arg && !has_crop_w_arg && !has_crop_d_arg,
+        "`crop` argument is not compatible with `crop_h`, `crop_w`, `crop_d`");
 
       auto cropArg = spec.GetRepeatedArgument<float>("crop");
       DALI_ENFORCE(cropArg.size() > 0 && cropArg.size() <= 2);
       crop_h = static_cast<int>(cropArg[0]);
-      crop_w = static_cast<int>(cropArg.size() == 2 ? cropArg[1] : cropArg[0]);
+      crop_w = static_cast<int>(cropArg.size() >= 2 ? cropArg[1] : cropArg[0]);
```
Comment (mzient) [inline dali/pipeline/operators/crop/crop_attr.h]:
```
    if (cropArg.size() == 3) {
      crop_d = cropArg[0];
      crop_h = cropArg[1];
      crop_w = cropArg[2];
    } else {
       // do it the old way
    }
```

Comment (jantonguirao) [reply dali/pipeline/operators/crop/crop_attr.h]: I am changing this in the next step PR #1210

**Pull request merged**